pennylane/hf/dipole.py

```diff
@@ -0,0 +1,310 @@
+
+    where the matrix elements :math:`d_{pq}` are given by the integral over molecular orbitals
+    :math:`\phi`
+
+    .. math::
+
+        d_{pq} = \int \phi_p^*(r) r \phi_q(r) dr,
+
+    :math:`c^{\dagger}` and :math:`c` are the creation and annihilation operators, respectively, and
+    :math:`D_n` is the contribution of the nuclei to the dipole operator.
+
+    The molecular orbitals are represented as a linear combination of atomic orbitals as
+
+    .. math::
+
+        \phi_i(r) = \sum_{\nu}c_{\nu}^i \chi_{\nu}(r).
+
+    Using this equation the dipole moment integral :math:`d_{pq}` can be written as
+
+    .. math::
+
+        d_{pq} = \sum_{\mu \nu} C_{p \mu} d_{\mu \nu} C_{\nu q},
+
+    where :math:`d_{\mu \nu}` is the dipole moment integral over the atomic orbitals and :math:`C`
+    is the molecular orbital expansion coefficient matrix.
+
+    Args:
+        mol (Molecule): the molecule object
+        core (list[int]): indices of the core orbitals
+        active (list[int]): indices of the active orbitals
+
+    Returns:
+        function: function that computes the dipole moment integrals in the molecular orbital basis
+
+    **Example**
+
+    >>> symbols  = ['H', 'H']
+    >>> geometry = np.array([[0.0, 0.0, 0.0], [0.0, 0.0, 1.0]], requires_grad = False)
+    >>> alpha = np.array([[3.42525091, 0.62391373, 0.1688554],
+    >>>                   [3.42525091, 0.62391373, 0.1688554]], requires_grad=True)
+    >>> mol = qml.hf.Molecule(symbols, geometry, alpha=alpha)
+    >>> args = [alpha]
+    >>> constants, integrals = dipole_integrals(mol)(*args)
+    >>> print(integrals)
+    (array([[0., 0.],
+            [0., 0.]]),
+     array([[0., 0.],
+            [0., 0.]]),
+     array([[ 0.5      , -0.8270995],
+            [-0.8270995,  0.5      ]]))
+    """
+
+    def _dipole_integrals(*args):
+        r"""Compute the dipole moment integrals in the molecular orbital basis.
+
+        Args:
+            args (array[array[float]]): initial values of the differentiable parameters
+
+        Returns:
+            tuple[array[float]]: tuple containing the core orbital contributions and the dipole
+            moment integrals
+        """
+        _, coeffs, _, _, _ = qml.hf.generate_scf(mol)(*args)
+
+        dx = anp.einsum("qr,rs,st->qt", coeffs.T, moment_matrix(mol.basis_set, 1, 0)(*args), coeffs)
+        dy = anp.einsum("qr,rs,st->qt", coeffs.T, moment_matrix(mol.basis_set, 1, 1)(*args), coeffs)
+        dz = anp.einsum("qr,rs,st->qt", coeffs.T, moment_matrix(mol.basis_set, 1, 2)(*args), coeffs)
+
+        core_x, core_y, core_z = anp.array([0]), anp.array([0]), anp.array([0])
+
+        if core is None and active is None:
+            return (core_x, core_y, core_z), (dx, dy, dz)
+
+        for i in core:
+            core_x = core_x + 2 * dx[i][i]
+            core_y = core_y + 2 * dy[i][i]
+            core_z = core_z + 2 * dz[i][i]
+
+        dx = dx[anp.ix_(active, active)]
```

Comment:
What does this `np.ix_` do?

Reply:
It takes the _active_ section of the whole dipole integral's matrix. (autograd might have complained with other methods to take that section out)